controllers/watcherapi_controller.go

```diff
@@ -18,42 +18,259 @@ package controllers
+
+	err = r.generateServiceConfigs(ctx, instance, secret, helper, &configVars)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	instance.Status.Conditions.MarkTrue(condition.ServiceConfigReadyCondition, condition.ServiceConfigReadyMessage)
+
+	// We reached the end of the Reconcile, update the Ready condition based on
+	// the sub conditions
+	if instance.Status.Conditions.AllSubConditionIsTrue() {
+		instance.Status.Conditions.MarkTrue(
+			condition.ReadyCondition, condition.ReadyMessage)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// generateServiceConfigs - create Secret which holds the service configuration
+// NOTE - jgilaber this function is WIP, currently implements a fraction of its
+// functionality and will be expanded of further iteration to actually generate
+// the service configs
+func (r *WatcherAPIReconciler) generateServiceConfigs(
+	ctx context.Context, instance *watcherv1beta1.WatcherAPI,
+	secret corev1.Secret,
+	helper *helper.Helper, envVars *map[string]env.Setter,
+) error {
+	Log := r.GetLogger(ctx)
+	Log.Info("generateServiceConfigs - reconciling")
+
+	db, err := mariadbv1.GetDatabaseByNameAndAccount(ctx, helper, watcher.DatabaseCRName, instance.Spec.DatabaseAccount, instance.Namespace)
+	if err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			condition.ServiceConfigReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			condition.ServiceConfigReadyErrorMessage,
+			err.Error()))
+		return err
+	}
+
+	// replace by actual usage in future iterations
+	_ = db
+	_ = secret
+	_ = envVars
+
+	return nil
+}
+
+func (r *WatcherAPIReconciler) reconcileDelete(ctx context.Context, instance *watcherv1beta1.WatcherAPI, helper *helper.Helper) (ctrl.Result, error) {
+	Log := r.GetLogger(ctx)
+	Log.Info(fmt.Sprintf("Reconcile Service '%s' delete started", instance.Name))
 
+	controllerutil.RemoveFinalizer(instance, helper.GetFinalizer())
+	Log.Info(fmt.Sprintf("Reconciled Service '%s' delete successfully", instance.Name))
 	return ctrl.Result{}, nil
 }
 
+func (r *WatcherAPIReconciler) initStatus(instance *watcherv1beta1.WatcherAPI) error {
+
+	cl := condition.CreateList(
+		// Mark ReadyCondition as Unknown from the beginning, because the
+		// Reconcile function is in progress. If this condition is not marked
+		// as True and is still in the "Unknown" state, we `Mirror(` the actual
+		// failure/in-progress operation
+		condition.UnknownCondition(condition.ReadyCondition, condition.InitReason, condition.ReadyInitMessage),
+		condition.UnknownCondition(condition.InputReadyCondition, condition.InitReason, condition.InputReadyInitMessage),
+		condition.UnknownCondition(condition.ServiceConfigReadyCondition, condition.InitReason, condition.ServiceConfigReadyMessage),
+	)
+
+	instance.Status.Conditions.Init(&cl)
+
+	// Update the lastObserved generation before evaluating conditions
+	instance.Status.ObservedGeneration = instance.Generation
```

Comment:
so on this pass I learned about generation and observedgeneration

does anyone know where we get the actual 'generation' from though? I mean, i understand this is incremented by the server itself on any changes in the spec, but i can't find it in the fields we define. We have observedGeneration added here but not generation. I expected to find it in the crd metadata.

Reply:
thanks for the link - and for confirming we don't see/carry it in our resource definitions (though 'metadata' exists on all resources including the ones we are adding here.)